README.md
# WirVsVirus Challenge - Team "LEIMdenC-API"

## Inspiration
Imagine you are a 
...mayor
...news agency
...researcher
...analyst
...anyone making decisions based on the data around Covid-19

and you need & want REALTIME access to all Covid-19 related facts <i>and<i> news for your area, focus group, age group.
But it is spread around, locked-up in dashboards, tables, charts or images.

## What it does
Built on the data from RKI Germany we provide an API that can be tailored to your needs. 
Built your query once - receive the currently reported data from RKI.
Compare here for correctness (correctness can in no way be guaranteed if data provisioning from RKI changes!!): https://experience.arcgis.com/experience/478220a4c454480e823b17327b2bf1d4

RKI Data gets usually refreshed 8:15am (https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/Fallzahlen.html), to allow for some room the API refreshes 9am.

## Try it out
A few examples:

* https://covid19-germany-api.eu-de.mybluemix.net/get_totals
* https://covid19-germany-api.eu-de.mybluemix.net/get_totals?state=Bayern
* https://covid19-germany-api.eu-de.mybluemix.net/get_totals?state=Bayern&date_range=2020-03-12 2020-03-21&province=SK München
* https://covid19-germany-api.eu-de.mybluemix.net/get_totals?state=Hamburg&date_range=2020-03-12 2020-03-21&sex=M
* https://covid19-germany-api.eu-de.mybluemix.net/get_data?state=Bayern&extraction_date=2020-03-24&province=SK München
* https://covid19-germany-api.eu-de.mybluemix.net/get_events?location=Thüringen
* https://covid19-germany-api.eu-de.mybluemix.net/get_events?date_range=2020-03-23%202020-03-27

Folgende weitere Parameter für /get_data & /get_totals sind verfügbar:

* state = Bundesland, alle 16 in Deutsch
* province = Landkreis, geführt mit SK(Stadtkreis) oder LK(Landkreis) - komplette Liste (Stand 23.03.2020) unten
* sex = Geschlecht
* age_group_start = Alter von
* age_group_end = Alter bis
* extraction_date = Abrufdatum in YYYY-MM-DD, jeden Tag wird ein kompletter Update gemacht und der alte erhalten, sodass auch Ältere Datensätze abgefragt werden können; weglassen setzt automatisch auf heute; erster Abruf am 26.03.2020
* date_range = Zeitraum von - bis anzugeben als YYYY-MM-DD YYYY-MM-DD (StartdatumLEERZEICHENEnddatum)
* province = Landkreis(province mit SK oder LK vorne), 
* sex = Geschlecht(M oder W), 
* age_group_start = Altersgrenze unten (beachte die Altersgruppen des RKI, nur innerhalb dieser wird gruppiert)
* age_group_end = Altersgrenze oben (beachte die Altersgruppen des RKI, nur innerhalb dieser wird gruppiert)


Folgende weitere Parameter für /get_events sind verfügbar, Datenbasis ist der Corono-Newsticker der Tagesschau:
* genannte Orte (location), Veröffentlichung (date_range YYYY-MM-DD YYYY-MM-DD), Abrufdatum (extraction_date), Artikelstichwortfilter (content), Titelstichwortfiler (title), genannte Organisationen (organization), genannte Personen (person)

## How we built it
Postgresql, Python Flask, Cloud Functions, Watson NLU, BeautifulSoup
![Intro](public/architecture.png)

## Challenges we ran into
Data, Data, Data ... the data is so hard to gather in a good quality.

## Accomplishments that we are proud of
We think that maaaany digital applications (whether its a news agency providing you daily charts or a political staff member providing the daily executive report for its official) can benefit from this. And while our solution can underpin many of those, we are proud that it is not only pure numbers we are providing but we also try to correlate this with events around the crisis on a local level.

## What We learned
My Hypothesis was: there is no easy access to large scale Covid-19 data
...turned out its true.
And some fun new technology and how to built an API and dynamic SQL :)

## What's next for CoVid-19-API
In the future we want to be able to aggregate news information on the COVID-19 virus from reliable sources while avoiding duplicates. Furthermore, we ourselves want to be able to publish some metrics/correlations from the data we gathered. At a later time, if the API is a success, we plan to have it reside inside a kubernetes cluster to ensure availability and scalability.

**Forks and merge requests are highly welcome!**

**Challenge ID** 0193	

**Schlüssel** 0038

Verfügbare Landkreise:
* 'SK Flensburg', 'SK Kiel', 'SK Lübeck', 'SK Neumünster',
       'LK Dithmarschen', 'LK Herzogtum Lauenburg', 'LK Nordfriesland',
       'LK Ostholstein', 'LK Pinneberg', 'LK Plön',
       'LK Rendsburg-Eckernförde', 'LK Schleswig-Flensburg',
       'LK Segeberg', 'LK Steinburg', 'LK Stormarn', 'SK Hamburg',
       'SK Braunschweig', 'SK Salzgitter', 'SK Wolfsburg', 'LK Gifhorn',
       'LK Goslar', 'LK Helmstedt', 'LK Northeim', 'LK Peine',
       'LK Wolfenbüttel', 'LK Göttingen', 'Region Hannover',
       'LK Diepholz', 'LK Hameln-Pyrmont', 'LK Hildesheim',
       'LK Holzminden', 'LK Nienburg (Weser)', 'LK Schaumburg',
       'LK Celle', 'LK Cuxhaven', 'LK Harburg', 'LK Lüchow-Dannenberg',
       'LK Lüneburg', 'LK Osterholz', 'LK Rotenburg (Wümme)',
       'LK Heidekreis', 'LK Stade', 'LK Uelzen', 'LK Verden',
       'SK Delmenhorst', 'SK Emden', 'SK Oldenburg', 'SK Osnabrück',
       'SK Wilhelmshaven', 'LK Ammerland', 'LK Aurich', 'LK Cloppenburg',
       'LK Emsland', 'LK Friesland', 'LK Grafschaft Bentheim', 'LK Leer',
       'LK Oldenburg', 'LK Osnabrück', 'LK Vechta', 'LK Wesermarsch',
       'LK Wittmund', 'SK Bremen', 'SK Bremerhaven', 'SK Düsseldorf',
       'SK Duisburg', 'SK Essen', 'SK Krefeld', 'SK Mönchengladbach',
       'SK Mülheim a.d.Ruhr', 'SK Oberhausen', 'SK Remscheid',
       'SK Solingen', 'SK Wuppertal', 'LK Kleve', 'LK Mettmann',
       'LK Rhein-Kreis Neuss', 'LK Viersen', 'LK Wesel', 'SK Bonn',
       'SK Köln', 'SK Leverkusen', 'StadtRegion Aachen', 'LK Düren',
       'LK Rhein-Erft-Kreis', 'LK Euskirchen', 'LK Heinsberg',
       'LK Oberbergischer Kreis', 'LK Rheinisch-Bergischer Kreis',
       'LK Rhein-Sieg-Kreis', 'SK Bottrop', 'SK Gelsenkirchen',
       'SK Münster', 'LK Borken', 'LK Coesfeld', 'LK Recklinghausen',
       'LK Steinfurt', 'LK Warendorf', 'SK Bielefeld', 'LK Gütersloh',
       'LK Herford', 'LK Höxter', 'LK Lippe', 'LK Minden-Lübbecke',
       'LK Paderborn', 'SK Bochum', 'SK Dortmund', 'SK Hagen', 'SK Hamm',
       'SK Herne', 'LK Ennepe-Ruhr-Kreis', 'LK Hochsauerlandkreis',
       'LK Märkischer Kreis', 'LK Olpe', 'LK Siegen-Wittgenstein',
       'LK Soest', 'LK Unna', 'SK Darmstadt', 'SK Frankfurt am Main',
       'SK Offenbach', 'SK Wiesbaden', 'LK Bergstraße',
       'LK Darmstadt-Dieburg', 'LK Groß-Gerau', 'LK Hochtaunuskreis',
       'LK Main-Kinzig-Kreis', 'LK Main-Taunus-Kreis', 'LK Odenwaldkreis',
       'LK Offenbach', 'LK Rheingau-Taunus-Kreis', 'LK Wetteraukreis',
       'LK Gießen', 'LK Lahn-Dill-Kreis', 'LK Limburg-Weilburg',
       'LK Marburg-Biedenkopf', 'LK Vogelsbergkreis', 'SK Kassel',
       'LK Fulda', 'LK Hersfeld-Rotenburg', 'LK Kassel',
       'LK Schwalm-Eder-Kreis', 'LK Waldeck-Frankenberg',
       'LK Werra-Meißner-Kreis', 'SK Koblenz', 'LK Ahrweiler',
       'LK Altenkirchen', 'LK Bad Kreuznach', 'LK Birkenfeld',
       'LK Cochem-Zell', 'LK Mayen-Koblenz', 'LK Neuwied',
       'LK Rhein-Hunsrück-Kreis', 'LK Rhein-Lahn-Kreis',
       'LK Westerwaldkreis', 'SK Trier', 'LK Bernkastel-Wittlich',
       'LK Bitburg-Prüm', 'LK Vulkaneifel', 'LK Trier-Saarburg',
       'SK Frankenthal', 'SK Kaiserslautern', 'SK Landau i.d.Pfalz',
       'SK Ludwigshafen', 'SK Mainz', 'SK Neustadt a.d.Weinstraße',
       'SK Pirmasens', 'SK Speyer', 'SK Worms', 'SK Zweibrücken',
       'LK Alzey-Worms', 'LK Bad Dürkheim', 'LK Donnersbergkreis',
       'LK Germersheim', 'LK Kaiserslautern', 'LK Kusel',
       'LK Südliche Weinstraße', 'LK Rhein-Pfalz-Kreis',
       'LK Mainz-Bingen', 'LK Südwestpfalz', 'SK Stuttgart',
       'LK Böblingen', 'LK Esslingen', 'LK Göppingen', 'LK Ludwigsburg',
       'LK Rems-Murr-Kreis', 'SK Heilbronn', 'LK Heilbronn',
       'LK Hohenlohekreis', 'LK Schwäbisch Hall', 'LK Main-Tauber-Kreis',
       'LK Heidenheim', 'LK Ostalbkreis', 'SK Baden-Baden',
       'SK Karlsruhe', 'LK Karlsruhe', 'LK Rastatt', 'SK Heidelberg',
       'SK Mannheim', 'LK Neckar-Odenwald-Kreis', 'LK Rhein-Neckar-Kreis',
       'SK Pforzheim', 'LK Calw', 'LK Enzkreis', 'LK Freudenstadt',
       'SK Freiburg i.Breisgau', 'LK Breisgau-Hochschwarzwald',
       'LK Emmendingen', 'LK Ortenaukreis', 'LK Rottweil',
       'LK Schwarzwald-Baar-Kreis', 'LK Tuttlingen', 'LK Konstanz',
       'LK Lörrach', 'LK Waldshut', 'LK Reutlingen', 'LK Tübingen',
       'LK Zollernalbkreis', 'SK Ulm', 'LK Alb-Donau-Kreis',
       'LK Biberach', 'LK Bodenseekreis', 'LK Ravensburg',
       'LK Sigmaringen', 'SK Ingolstadt', 'SK München', 'SK Rosenheim',
       'LK Altötting', 'LK Berchtesgadener Land',
       'LK Bad Tölz-Wolfratshausen', 'LK Dachau', 'LK Ebersberg',
       'LK Eichstätt', 'LK Erding', 'LK Freising', 'LK Fürstenfeldbruck',
       'LK Garmisch-Partenkirchen', 'LK Landsberg a.Lech', 'LK Miesbach',
       'LK Mühldorf a.Inn', 'LK München', 'LK Neuburg-Schrobenhausen',
       'LK Pfaffenhofen a.d.Ilm', 'LK Rosenheim', 'LK Starnberg',
       'LK Traunstein', 'LK Weilheim-Schongau', 'SK Landshut',
       'SK Passau', 'SK Straubing', 'LK Deggendorf',
       'LK Freyung-Grafenau', 'LK Kelheim', 'LK Landshut', 'LK Passau',
       'LK Regen', 'LK Rottal-Inn', 'LK Straubing-Bogen',
       'LK Dingolfing-Landau', 'SK Amberg', 'SK Regensburg',
       'SK Weiden i.d.OPf.', 'LK Amberg-Sulzbach', 'LK Cham',
       'LK Neumarkt i.d.OPf.', 'LK Neustadt a.d.Waldnaab',
       'LK Regensburg', 'LK Tirschenreuth', 'SK Bamberg', 'SK Bayreuth',
       'SK Hof', 'LK Bamberg', 'LK Bayreuth', 'LK Coburg', 'LK Forchheim',
       'LK Hof', 'LK Kronach', 'LK Kulmbach', 'LK Lichtenfels',
       'LK Wunsiedel i.Fichtelgebirge', 'SK Ansbach', 'SK Erlangen',
       'SK Fürth', 'SK Nürnberg', 'SK Schwabach', 'LK Ansbach',
       'LK Erlangen-Höchstadt', 'LK Fürth', 'LK Nürnberger Land',
       'LK Neustadt a.d.Aisch-Bad Windsheim', 'LK Roth',
       'LK Weißenburg-Gunzenhausen', 'SK Aschaffenburg', 'SK Schweinfurt',
       'SK Würzburg', 'LK Aschaffenburg', 'LK Bad Kissingen',
       'LK Rhön-Grabfeld', 'LK Haßberge', 'LK Kitzingen', 'LK Miltenberg',
       'LK Main-Spessart', 'LK Schweinfurt', 'LK Würzburg', 'SK Augsburg',
       'SK Memmingen', 'LK Aichach-Friedberg', 'LK Augsburg',
       'LK Dillingen a.d.Donau', 'LK Günzburg', 'LK Neu-Ulm', 'LK Lindau',
       'LK Ostallgäu', 'LK Unterallgäu', 'LK Donau-Ries', 'LK Oberallgäu',
       'LK Stadtverband Saarbrücken', 'LK Merzig-Wadern',
       'LK Neunkirchen', 'LK Saarlouis', 'LK Saar-Pfalz-Kreis',
       'LK Sankt Wendel', 'SK Berlin Mitte',
       'SK Berlin Friedrichshain-Kreuzberg', 'SK Berlin Pankow',
       'SK Berlin Charlottenburg-Wilmersdorf', 'SK Berlin Spandau',
       'SK Berlin Steglitz-Zehlendorf', 'SK Berlin Tempelhof-Schöneberg',
       'SK Berlin Neukölln', 'SK Berlin Treptow-Köpenick',
       'SK Berlin Marzahn-Hellersdorf', 'SK Berlin Lichtenberg',
       'SK Berlin Reinickendorf', 'SK Brandenburg a.d.Havel',
       'SK Cottbus', 'SK Frankfurt (Oder)', 'SK Potsdam', 'LK Barnim',
       'LK Dahme-Spreewald', 'LK Elbe-Elster', 'LK Havelland',
       'LK Märkisch-Oderland', 'LK Oberhavel', 'LK Oberspreewald-Lausitz',
       'LK Oder-Spree', 'LK Ostprignitz-Ruppin', 'LK Potsdam-Mittelmark',
       'LK Spree-Neiße', 'LK Teltow-Fläming', 'LK Uckermark',
       'SK Rostock', 'SK Schwerin', 'LK Mecklenburgische Seenplatte',
       'LK Rostock', 'LK Vorpommern-Rügen', 'LK Nordwestmecklenburg',
       'LK Vorpommern-Greifswald', 'LK Ludwigslust-Parchim',
       'SK Chemnitz', 'LK Erzgebirgskreis', 'LK Mittelsachsen',
       'LK Vogtlandkreis', 'LK Zwickau', 'SK Dresden', 'LK Bautzen',
       'LK Görlitz', 'LK Meißen', 'LK Sächsische Schweiz-Osterzgebirge',
       'SK Leipzig', 'LK Leipzig', 'LK Nordsachsen', 'SK Dessau-Roßlau',
       'SK Halle', 'SK Magdeburg', 'LK Altmarkkreis Salzwedel',
       'LK Anhalt-Bitterfeld', 'LK Börde', 'LK Burgenlandkreis',
       'LK Harz', 'LK Jerichower Land', 'LK Mansfeld-Südharz',
       'LK Saalekreis', 'LK Salzlandkreis', 'LK Stendal', 'LK Wittenberg',
       'SK Erfurt', 'SK Gera', 'SK Jena', 'SK Suhl', 'SK Weimar',
       'SK Eisenach', 'LK Eichsfeld', 'LK Nordhausen', 'LK Wartburgkreis',
       'LK Unstrut-Hainich-Kreis', 'LK Kyffhäuserkreis',
       'LK Schmalkalden-Meiningen', 'LK Gotha', 'LK Sömmerda',
       'LK Hildburghausen', 'LK Ilm-Kreis', 'LK Weimarer Land',
       'LK Sonneberg', 'LK Saalfeld-Rudolstadt',
       'LK Saale-Holzland-Kreis', 'LK Saale-Orla-Kreis', 'LK Greiz',
       'LK Altenburger Land'
       
## License

This application is licensed under the Apache License, Version 2. Separate third-party code objects invoked within this code pattern are licensed by their respective providers pursuant to their own separate licenses. Contributions are subject to the [Developer Certificate of Origin, Version 1.1](https://developercertificate.org/) and the [Apache License, Version 2](https://www.apache.org/licenses/LICENSE-2.0.txt).

[Apache License FAQ](https://www.apache.org/foundation/license-faq.html#WhatDoesItMEAN)



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
* ansible playbook that provisions a cron-job that runs every midnight and should trigger aggregator
and routines

[Unreleased]: https://github.com/maxisses/Covid-19API/compare/feature/add-aggregation-cron-job?expand=1

flaskAPIwDocker/hello/app.py
import os
from flask import Flask
from flask import jsonify
from flask import request
from flask import render_template
from datetime import datetime, timedelta
import psycopg2
import psycopg2.extras

app = Flask(__name__)

def create_app():
    """
    Create a Flask application using the app factory pattern.
    :param settings_override: Override settings
    :return: Flask app
    """
    app = Flask(__name__)

    cases = [
            {  
                "country": "Germany",
                "date": "19.03.2020",
                "lat": "51",
                "long": "9",
                "total_cases": 12000,
                "active_cases": 11000,
                "died_cases": 50
            },
            {  
                "country": "Germany",
                "date": "18.03.2020",
                "lat": "51",
                "long": "9",
                "total_cases": 11500,
                "active_cases": 11000,
                "died_cases": 30
            },
            {  
                "country": "France",
                "date": "19.03.2020",
                "lat": "46.2276",
                "long": "2.2137",
                "total_cases": 15000,
                "active_cases": 13000,
                "died_cases": 60
            }
        ]

    """ 
    mögliche Endpunkte der technischen Schnittstelle (API):

    Zahlen:

    get_cases (parameter = zeitraum, nation, bundesland, city, gender, agegroup ... )
    --> haben wir nich; get_recovered_per_day ( -,,-)
    --> haben wir nich; get_all_recoverd ( -,,-)
    get_active_cases ( -,,-)
    get_prediction ( -,,-)
    --> basierend auf Regression oder mehr

    Maßnahmen:

    get_political_measures (returns = datum, name, beschreibung, ... ) (parameter = Zeitraum, nation, ...)
    get_medical_measures (returns = name, beschreibung)
    --> im Kontext Ibuprofen eingeschränkt, in die Zukunft: impstoff gefunden, etc.
    """


    @app.route('/')
    def index():
        """
        Render our API documentation and some graphics here?
        """
        return render_template("index.html")

    # GET /get_total_infected
    # (parameter = zeitraum, nation, bundesland, city, gender, agegroup ... )
    @app.route("/get_totals")
    def get_totals():
        state = request.args.get('state')
        province = request.args.get('province')
        sex = request.args.get('sex')
        age_group_start = request.args.get('age_group_start')

        try:
            age_group_start = int(age_group_start)
        except:
            pass
        age_group_end = request.args.get('age_group_end')

        try:
            age_group_end = int(age_group_end)
        except:
            pass

        extraction_date = request.args.get('extraction_date')
        
        extraction_date = handle_extraction_date(extraction_date)

        date_range = request.args.get('date_range')

        date_range_start = None
        date_range_end = None
        try:
            date_range = date_range.split(" ")
            print("fetching the diff between two dates: " + " ".join(date_range))
            if len(date_range) > 1:
                date_range_start = date_range[0]
                date_range_end = date_range[1]
            else:
                date_range_start = date_range[0]
                date_range_end = date_range[0]
        except:
            pass

        dbname = os.environ['DB_DATABASE']
        user = os.environ['DB_USER']
        password = os.environ['DB_PASSWORD']
        host = os.environ['DB_HOST']

        print(dbname, user, password, host)

        conn = psycopg2.connect("dbname='wirvsvirus' user='wirvsvirus' host='marc-book.de' password='[n2^3kKCyxUGgzuV'")
        cur = conn.cursor(cursor_factory = psycopg2.extras.RealDictCursor)

        ## conn = psycopg2.connect("dbname='{}' user='{}' host='{}'".format(dbname, user, host))
        ## cur = conn.cursor(cursor_factory = psycopg2.extras.RealDictCursor)

        selection = {"extraction_date": extraction_date,"state":state, "province":province, "sex": sex, 
                    "age_group_start":age_group_start, "age_group_end": age_group_end}

        columns = []
        for key, value in selection.items():
            if value:
                columns.append(key)
        columns = ", ".join(columns)
        print(columns)

        
        selection["date_range_start"] = date_range_start
        selection["date_range_end"] = date_range_end

                ### get the numbers for the end
        cur.execute(""" SELECT {} , sum(case_count) AS infected, sum(death_count) AS deceased
                        FROM rki_data_germany 
                            WHERE (extraction_date = %(extraction_date)s OR %(extraction_date)s IS NULL )
                            AND (state = %(state)s OR %(state)s IS NULL )
                            AND (province = %(province)s OR %(province)s IS NULL)
                            AND (sex = %(sex)s OR %(sex)s IS NULL)
                            AND (age_group_start >= %(age_group_start)s OR %(age_group_start)s IS NULL)
                            AND (age_group_end <= %(age_group_end)s OR %(age_group_end)s IS NULL)
                            AND ((notification_date <= %(date_range_end)s) OR %(date_range_end)s IS NULL)
                            GROUP BY {}""".format(columns, columns), 
                            selection)

        rows_end = cur.fetchall()

        ### get the numbers for the start date
        cur.execute(""" SELECT {} , sum(case_count) AS infected, sum(death_count) AS deceased
                        FROM rki_data_germany 
                            WHERE (extraction_date = %(extraction_date)s OR %(extraction_date)s IS NULL )
                            AND (state = %(state)s OR %(state)s IS NULL )
                            AND (province = %(province)s OR %(province)s IS NULL)
                            AND (sex = %(sex)s OR %(sex)s IS NULL)
                            AND (age_group_start >= %(age_group_start)s OR %(age_group_start)s IS NULL)
                            AND (age_group_end <= %(age_group_end)s OR %(age_group_end)s IS NULL)
                            AND ((notification_date <= %(date_range_start)s) OR %(date_range_start)s IS NULL)
                            GROUP BY {}""".format(columns, columns), 
                            selection)

        rows_start = cur.fetchall()

        cur.execute(""" SELECT {} , sum(case_count) AS infected, sum(death_count) AS deceased
                        FROM rki_data_germany 
                            WHERE (extraction_date = %(extraction_date)s OR %(extraction_date)s IS NULL )
                            AND (state = %(state)s OR %(state)s IS NULL )
                            AND (province = %(province)s OR %(province)s IS NULL)
                            AND (sex = %(sex)s OR %(sex)s IS NULL)
                            AND (age_group_start >= %(age_group_start)s OR %(age_group_start)s IS NULL)
                            AND (age_group_end <= %(age_group_end)s OR %(age_group_end)s IS NULL)
                            AND ((notification_date BETWEEN %(date_range_start)s AND %(date_range_end)s) OR %(date_range_start)s IS NULL)
                            GROUP BY {}""".format(columns, columns), 
                            selection)

        rows = cur.fetchall()

        if len(rows) == 0:
            return jsonify({"message": "error or no values"})
        else:
            if date_range_start:
                print(rows_end[0]["infected"])
                print(rows_start[0]["infected"])
                rows[0]["diff_infected"] =  rows_end[0]["infected"] - rows_start[0]["infected"]
                rows[0]["diff_deceased"] =  rows_end[0]["deceased"] - rows_start[0]["deceased"]
                rows[0]["from"] = date_range_start
                rows[0]["to"] = date_range_end
            return jsonify(rows)

    # GET /get_data
    @app.route("/get_data")
    def get_data():
        state = request.args.get('state')
        province = request.args.get('province')
        sex = request.args.get('sex')
        age_group_start = request.args.get('age_group_start')

        try:
            age_group_start = int(age_group_start)
        except:
            pass
        age_group_end = request.args.get('age_group_end')

        try:
            age_group_end = int(age_group_end)
        except:
            pass
        extraction_date = request.args.get('extraction_date')

        extraction_date = handle_extraction_date(extraction_date)

        date_range = request.args.get('date_range')

        date_range_start = None
        date_range_end = None
        try:
            date_range = date_range.split(" ")
            print(date_range)
            if len(date_range) > 1:
                date_range_start = date_range[0]
                date_range_end = date_range[1]
            else:
                date_range_start = date_range[0]
                date_range_end = date_range[0]
        except:
            pass

        conn = psycopg2.connect("dbname='wirvsvirus' user='wirvsvirus' host='marc-book.de' password='[n2^3kKCyxUGgzuV'")
        cur = conn.cursor(cursor_factory = psycopg2.extras.RealDictCursor)

        selection = {"extraction_date": extraction_date,"state":state, "province":province, "sex": sex, 
                    "age_group_start":age_group_start, "age_group_end": age_group_end}

        columns = []
        for key, value in selection.items():
            columns.append(key)
        columns = ", ".join(columns)
        print(columns)
        
        selection["date_range_start"] = date_range_start
        selection["date_range_end"] = date_range_end

        print(selection)
        cur.execute(""" SELECT {} ,notification_date AS reported, case_count AS infected, death_count AS deceased
                        FROM rki_data_germany 
                            WHERE (extraction_date = %(extraction_date)s OR %(extraction_date)s IS NULL )
                            AND (state = %(state)s OR %(state)s IS NULL )
                            AND (province = %(province)s OR %(province)s IS NULL)
                            AND (sex = %(sex)s OR %(sex)s IS NULL)
                            AND (age_group_start >= %(age_group_start)s OR %(age_group_start)s IS NULL)
                            AND (age_group_end <= %(age_group_end)s OR %(age_group_end)s IS NULL)
                            AND ((notification_date BETWEEN %(date_range_start)s AND %(date_range_end)s) OR %(date_range_start)s IS NULL)
                            """.format(columns, columns), 
                            selection)

        rows = cur.fetchall()

        if len(rows) == 0:
            return jsonify({"message": "error or no values"})
        else:
            if date_range_start:
                for row in rows:
                    row["from"] = date_range_start
                    row["to"] = date_range_end
            return jsonify(rows)

    # GET /get_events
    @app.route("/get_events")
    def get_events():
        location = request.args.get('location')
        referred_date = request.args.get('publish_date')
        extraction_date = request.args.get('extraction_date')

        conn = psycopg2.connect("dbname='wirvsvirus' user='wirvsvirus' host='marc-book.de' password='[n2^3kKCyxUGgzuV'")
        cur = conn.cursor(cursor_factory = psycopg2.extras.RealDictCursor)

        selection = {"extraction_date": extraction_date, "referred_date": referred_date, "location": location}

        columns = []
        for key, value in selection.items():
            columns.append(key)
        columns = ", ".join(columns)
        print(columns)

        print(selection)
        cur.execute(""" SELECT *
                        FROM corona_events
                            WHERE (extraction_date = %(extraction_date)s OR %(extraction_date)s IS NULL )
                            AND (referred_date = %(referred_date)s OR %(referred_date)s IS NULL )
                            AND (event_location = %(location)s OR %(location)s IS NULL)
                            """, 
                            selection)

        rows = cur.fetchall()

        if len(rows) == 0:
            return jsonify({"message": "error or no values"})
        else:
            return jsonify(rows)


    #############################

    def handle_extraction_date(date):
        if date:
            return date
        else:
            #extraction_date = (datetime.now() - timedelta(1)).strftime('%Y-%m-%d')
            date = "2020-03-21"
            print("set extraction date to yesterday: " + str(date))
            return date
    
    return app
